File: examples/c/basic_usage.c
/**
 * Basic usage example for libpsam
 *
 * Demonstrates:
 * - Creating a model
 * - Training on token sequences
 * - Making predictions
 * - Saving and loading
 */

#include <stdio.h>
#include <stdlib.h>
#include <psam.h>

int main() {
    printf("╔════════════════════════════════════════════════════════════╗\n");
    printf("║          libpsam - Basic Usage Example                    ║\n");
    printf("╚════════════════════════════════════════════════════════════╝\n\n");

    // Configuration
    const uint32_t VOCAB_SIZE = 100;
    const uint32_t WINDOW = 8;
    const uint32_t TOP_K = 10;

    // Create model
    printf("📦 Creating PSAM model...\n");
    printf("   - Vocabulary size: %u\n", VOCAB_SIZE);
    printf("   - Window: %u\n", WINDOW);
    printf("   - Top-K: %u\n\n", TOP_K);

    psam_model_t* model = psam_create(VOCAB_SIZE, WINDOW, TOP_K);
    if (!model) {
        fprintf(stderr, "❌ Failed to create model\n");
        return 1;
    }

    // Training data: "the quick brown fox jumps over the lazy dog"
    uint32_t tokens[] = {1, 2, 3, 4, 5, 6, 1, 7, 8};
    size_t num_tokens = sizeof(tokens) / sizeof(tokens[0]);

    printf("📚 Training on sequence...\n");
    printf("   Tokens: [");
    for (size_t i = 0; i < num_tokens; i++) {
        printf("%u%s", tokens[i], i < num_tokens - 1 ? ", " : "");
    }
    printf("]\n\n");

    // Train
    psam_error_t err = psam_train_batch(model, tokens, num_tokens);
    if (err != PSAM_OK) {
        fprintf(stderr, "❌ Training failed: %s\n", psam_error_string(err));
        psam_destroy(model);
        return 1;
    }

    // Finalize training
    err = psam_finalize_training(model);
    if (err != PSAM_OK) {
        fprintf(stderr, "❌ Finalization failed: %s\n", psam_error_string(err));
        psam_destroy(model);
        return 1;
    }

    printf("✓ Training complete!\n\n");

    // Get statistics
    psam_stats_t stats;
    err = psam_get_stats(model, &stats);
    if (err == PSAM_OK) {
        printf("📊 Model Statistics:\n");
        printf("   - Vocabulary: %u tokens\n", stats.vocab_size);
        printf("   - Rows: %u\n", stats.row_count);
        printf("   - Edges: %llu\n", (unsigned long long)stats.edge_count);
        printf("   - Memory: %llu bytes (%.1f KB)\n\n",
               (unsigned long long)stats.memory_bytes,
               stats.memory_bytes / 1024.0);
    }

    // Make predictions
    printf("🔮 Making predictions...\n");
    uint32_t context[] = {1, 2, 3};  // "the quick brown"
    size_t context_len = sizeof(context) / sizeof(context[0]);

    printf("   Context: [");
    for (size_t i = 0; i < context_len; i++) {
        printf("%u%s", context[i], i < context_len - 1 ? ", " : "");
    }
    printf("]\n\n");

    psam_prediction_t predictions[TOP_K];
    int num_preds = psam_predict(model, context, context_len, predictions, TOP_K);

    if (num_preds < 0) {
        fprintf(stderr, "❌ Prediction failed: %s\n",
                psam_error_string((psam_error_t)num_preds));
    } else {
        printf("   Predictions:\n");
        for (int i = 0; i < num_preds; i++) {
            printf("   %2d. Token %u (score: %.3f)\n",
                   i + 1, predictions[i].token_id, predictions[i].score);
        }
        printf("\n");
    }

    // Save model
    const char* filepath = "example_model.psam";
    printf("💾 Saving model to '%s'...\n", filepath);

    err = psam_save(model, filepath);
    if (err != PSAM_OK) {
        fprintf(stderr, "❌ Save failed: %s\n", psam_error_string(err));
    } else {
        printf("✓ Model saved!\n\n");
    }

    // Clean up
    psam_destroy(model);

    // Load model back
    printf("📂 Loading model from '%s'...\n", filepath);
    psam_model_t* loaded = psam_load(filepath);

    if (!loaded) {
        fprintf(stderr, "❌ Load failed\n");
        return 1;
    }

    printf("✓ Model loaded!\n\n");

    // Verify loaded model works
    printf("🔮 Testing loaded model...\n");
    num_preds = psam_predict(loaded, context, context_len, predictions, TOP_K);

    if (num_preds > 0) {
        printf("✓ Loaded model works! First prediction: Token %u\n\n",
               predictions[0].token_id);
    }

    // Cleanup
    psam_destroy(loaded);

    printf("🎉 Example complete!\n");
    printf("\n");
    printf("libpsam version: %s\n", psam_version());

    return 0;
}

File: examples/python/basic_usage.py
#!/usr/bin/env python3
"""
Basic usage example for libpsam

Demonstrates:
- Creating a model
- Training on token sequences
- Making predictions
- Saving and loading
"""

from psam import PSAM, is_library_available

def main():
    print("╔════════════════════════════════════════════════════════════╗")
    print("║          libpsam - Python Basic Usage Example             ║")
    print("╚════════════════════════════════════════════════════════════╝\n")

    # Check if native library is available
    if not is_library_available():
        print("❌ Native library not available")
        print("   Build libpsam.so and set LIBPSAM_PATH environment variable")
        return 1

    print("✓ Native library loaded\n")

    # Configuration
    VOCAB_SIZE = 100
    WINDOW = 8
    TOP_K = 10

    print("📦 Creating PSAM model...")
    print(f"   - Vocabulary size: {VOCAB_SIZE}")
    print(f"   - Window: {WINDOW}")
    print(f"   - Top-K: {TOP_K}\n")

    psam = PSAM(vocab_size=VOCAB_SIZE, window=WINDOW, top_k=TOP_K)

    # Training data: "the quick brown fox jumps over the lazy dog"
    tokens = [1, 2, 3, 4, 5, 6, 1, 7, 8]

    print("📚 Training on sequence...")
    print(f"   Tokens: {tokens}\n")

    psam.train_batch(tokens)
    psam.finalize_training()

    print("✓ Training complete!\n")

    # Get statistics
    stats = psam.stats()
    print("📊 Model Statistics:")
    print(f"   - Vocabulary: {stats.vocab_size} tokens")
    print(f"   - Rows: {stats.row_count}")
    print(f"   - Edges: {stats.edge_count}")
    print(f"   - Memory: {stats.memory_bytes} bytes ({stats.memory_bytes / 1024:.1f} KB)\n")

    # Make predictions
    print("🔮 Making predictions...")
    context = [1, 2, 3]  # "the quick brown"
    print(f"   Context: {context}\n")

    token_ids, scores = psam.predict(context, max_predictions=5)

    print("   Predictions:")
    for i, (token_id, score) in enumerate(zip(token_ids, scores)):
        print(f"   {i + 1}. Token {token_id} (score: {score:.3f})")
    print()

    # Sample from distribution
    print("🎲 Sampling with different temperatures...")
    for temp in [0.5, 1.0, 2.0]:
        sampled = psam.sample(context, temperature=temp)
        print(f"   T={temp}: Token {sampled}")
    print()

    # Save model
    filepath = "example_model.psam"
    print(f"💾 Saving model to '{filepath}'...")

    psam.save(filepath)
    print("✓ Model saved!\n")

    # Load model back
    print(f"📂 Loading model from '{filepath}'...")
    loaded = PSAM.load(filepath)
    print("✓ Model loaded!\n")

    # Verify loaded model works
    print("🔮 Testing loaded model...")
    loaded_ids, loaded_scores = loaded.predict(context, max_predictions=3)
    print(f"✓ Loaded model works! First prediction: Token {loaded_ids[0]}\n")

    # Cleanup
    psam.destroy()
    loaded.destroy()

    print("🎉 Example complete!\n")
    print(f"libpsam version: {PSAM.version()}")

    return 0

if __name__ == "__main__":
    exit(main())
